Post transformation validation

cvalarida opened the pull request:
## Description
So we can ensure the resulting data structure is what we expect from the JSON schema.

I ended up adding the `contentModelType` onto the entity because we weren't able to determine the post-transformation entity type with the properties we had. So now we have three similar properties:
- `baseType` to help find the file to read
- `contentModelType` to do content model type things (lots)
- `entityBundle` which will be specified by each transformer (at least that's what we're doing now) and is used in the templates

## Testing done
All manual.

## Acceptance criteria
- [ ] Unit tests fail when validation errors are detected
- [ ] Post-transformation JSON schema validation is run

Commit:
Create validateRawEntity and validateTransformedEntity

Commit:
Move node-page raw JSON schema

Commit:
Validate after transformation

Commit:
Add contentModelType to read entity

Commit:
Update testing the first node

Commit:
Use $id in the schemas

Commit:
Thorw errors to fail in unit tests; log the invalid data

Commit:
Add all the schema files to ajv validator

Commit:
Add some basic schema files

Commit:
Add transformed node-page schema

Commit:
Add generic Paragraph schema

Commit:
Add nullable-string

Commit:
Remove nullable-string schema

Because we can specify multiple schemas, turns out

Commit:
Fix up transformed node-page schema

Commit:
Add properties to paragraph schemas

Commit:
Use anyOf instead of oneOf in Paragraph schema

Commit:
Pull out meta tags schema

Commit:
Merge branch 'master' into post-transformation-validation-cv

Commit:
Update transformed node-page entity test json

cvalarida commented on src/site/stages/build/process-cms-exports/schema-validation.js:
```diff
@@ -7,47 +7,56 @@ const validate = require('./validator');
+const rawSchemasDir = path.join(__dirname, 'schemas', 'raw');
+const transformedSchemasDir = path.join(__dirname, 'schemas', 'transformed');
+
+const validateEntityFactory = schemasDir => {
```
This is so we can have a validator for raw and transformed schemas separately. They'll need to keep track of their own missing schemas, so I wrapped that up in here as well.

cvalarida commented on src/site/stages/build/process-cms-exports/schemas/transformed/node-page.js:
```diff
@@ -0,0 +1,56 @@
+      $ref: 'MetaTags',
+    },
+  },
+  // required: getFilter('node-page'),
```
We'll probably want to specify all the fields that are actually used in the templates.

kelsonic replied:
Standard key-value pairs for Node entities:
```
- `changed`
- `created`
- `default_langcode`
- `langcode`
- `menu_link`
- `moderation_state`
- `path`
- `promote`
- `revision_log`
- `revision_timestamp`
- `revision_translation_affected`
- `revision_uid`
- `status`
- `sticky`
- `title`
- `type`
- `uid`
- `uuid`
```

Key-value pairs for `node` `page`s:
```
  - `field_administration`
  - `field_alert`
  - `field_content_block`
  - `field_description`
  - `field_featured_content`
  - `field_intro_text`
  - `field_meta_tags`
  - `field_meta_title`
  - `field_page_last_built`
  - `field_plainlanguage_date`
  - `field_related_links`
```

From `src/site/stages/build/process-cms-exports/graphql-cms-export-differences/node.md` :)

cvalarida commented on src/site/stages/build/process-cms-exports/schemas/transformed/node-page.js:
```diff
@@ -0,0 +1,56 @@
+// const { getFilter } = require('../../filters');
+
+module.exports = {
+  $id: 'node-page',
```
Each schema will need an `$id` so we can use it in a `$ref`.

cvalarida commented on src/site/stages/build/process-cms-exports/validator.js:
```diff
@@ -1,20 +1,25 @@
+/**
+ * Loads all the schemas in a directory into ajv
+ */
+const loadSchemas = dir => {
```
With this, we won't have to add schemas individually whenever we need them in a `$ref`.

cvalarida commented on src/site/stages/build/process-cms-exports/filters.js:
```diff
@@ -5,7 +5,7 @@ const { getContentModelType } = require('./helpers');
 
 const whitelists = {
-  global: ['title'],
+  global: ['title', 'baseType', 'contentModelType'],
```
We need these in the transformed entities.

...:thinking: Or at least `contentModelType`...

bkjohnson commented on src/site/stages/build/process-cms-exports/schemas/transformed/paragraph-list-of-link-teasers.js:
```diff
@@ -0,0 +1,12 @@
+  $id: contentModelType,
+  type: 'object',
+  baseType: {
+    enum: ['paragraph'],
```
Is `properties` intentionally left out as a wrapper for `baseType` and `contentModelType`?

cvalarida replied:
So these differ only a little for now because they're incomplete. I just needed something to refer to from the `node-page.js` schema. Once we get around to filling them in, we can decide whether it makes sense to pull some common elements together in one place.

bkjohnson replied:
Gotcha :+1:

bkjohnson commented on src/site/stages/build/process-cms-exports/schemas/transformed/paragraph-q-a.js:
```diff
@@ -0,0 +1,14 @@
+  type: 'object',
+  properties: {
+    baseType: {
+      enum: ['paragraph'],
```
here

bkjohnson commented on src/site/stages/build/process-cms-exports/schemas/transformed/taxonomy-term-administration.js:
```diff
@@ -0,0 +1,14 @@
+  type: 'object',
+  properties: {
+    baseType: {
+      enum: ['taxonomy_term'],
```
here

bkjohnson reviewed (commented):
I noticed that several of the schemas differ only in their `baseType` and `contentModelType`.  Would it be useful to put something into `schemas/common` that lets us call something like `genericTypedObject(contentModelType, baseType)` and have the output of that be the export?

bkjohnson reviewed (approved):
:+1: looking forward to validating some transformations!

Commit:
Add logging because the objects are too large for Mocha to show the diff

Commit:
Add newline

Commit:
Better error messaging for circular reference error

Commit:
Throw error instead of process.exit(1)

Commit:
Use ignoreList to ignore not just types

Some entity references are circular. Before we create the filters for
the individual content models, we need a way to exclude
them.

Commit:
Attempt getting the UTC time right

Commit:
fieldPageLastBuilt is { date: 'string' } not 'string'

Commit:
Use moment-timezone

Commit:
Update test data

Commit:
Add comment

cvalarida commented on src/site/stages/build/process-cms-exports/tests/assemble-entity-tree.unit.spec.js:
```diff
@@ -71,7 +71,23 @@ describe('CMS export', () => {
+              'Transformed from assembleEntityTree:\n',
+              JSON.stringify(assembled, null, 2),
+            );
+            throw e;
```
Turns out without this extra logging, it won't log what's different when they're very large objects, so we can't tell what the differences are.

Commit:
Rename transformed schemas

Commit:
Dynamically name schemas

Commit:
Remove $id from schemas

Commit:
Update schemas to reference namespaced schemas

Commit:
Add comment for dynamic schema naming

cvalarida commented on src/site/stages/build/process-cms-exports/validator.js:
```diff
@@ -1,20 +1,31 @@
+      schema,
+      // Common schemas should have $id, but dynamically name raw and transformed schemas
+      schema.$id ||
+        `${dir.split(path.sep).pop()}/${path.basename(fileName, '.js')}`,
```
Alternatively, we can pass in two parameters—one for the path, and one for the namespace. I'm fine either way. `dir.split(path.sep).pop()` is kinda gross, but then again, so is passing in more parameters than we need. :man_shrugging:

bkjohnson replied:
It isn't _super_ obvious that this is supposed to be `raw/[entityType]` or `transformed/[entityType]`, but I think it's ok for the time being.  It think it's more useful to have this in so we can have the validation on our upcoming transformer PRs

bkjohnson commented on src/site/stages/build/process-cms-exports/index.js:
```diff
@@ -1,10 +1,40 @@
+ * expansion of certain properties before we've created a filter for
+ * that content model.
+ */
+const ignoreList = typeProperties.concat([
```
So is the idea that we will remove things from this as we go?

cvalarida replied:
Yeah, exactly. There may end up being things we keep in there, but for the time being, this is like a temporary list until we write the filters for the individual content types which cause the circular references and whatnot.

bkjohnson reviewed (approved):
Small questions but no objections.
